foreman/data_refinery_foreman/foreman/test_main.py

```diff
@@ -0,0 +1,274 @@
+        batch = self.insert_batch()
+        job = DownloaderJob.create_job_and_relationships(
+            num_retries=0, batches=[batch], downloader_task="dummy_task")
+        job.success = False
+        job.save()
+
+        # Just run it once, not forever so get the function that is
+        # decorated with @do_forever
+        main.retry_failed_downloader_jobs.__wrapped__()
+        mock_send_task.assert_called_once()
+
+        jobs = DownloaderJob.objects.order_by('id')
+        original_job = jobs[0]
+        self.assertTrue(original_job.retried)
+        self.assertEqual(original_job.num_retries, 0)
+        self.assertFalse(original_job.success)
+
+        retried_job = jobs[1]
+        self.assertEqual(retried_job.num_retries, 1)
+
+    @patch('data_refinery_foreman.surveyor.message_queue.app.send_task')
+    def test_retrying_hung_downloader_jobs(self, mock_send_task):
+        batch = self.insert_batch()
+        job = DownloaderJob.create_job_and_relationships(
+            num_retries=0, batches=[batch], downloader_task="dummy_task")
+        job.start_time = timezone.now() - main.MAX_DOWNLOADER_RUN_TIME - timedelta(seconds=1)
+        job.save()
+
+        # Just run it once, not forever so get the function that is
+        # decorated with @do_forever
+        main.retry_hung_downloader_jobs.__wrapped__()
+        mock_send_task.assert_called_once()
+
+        jobs = DownloaderJob.objects.order_by('id')
+        original_job = jobs[0]
+        self.assertTrue(original_job.retried)
+        self.assertEqual(original_job.num_retries, 0)
+        self.assertFalse(original_job.success)
+
+        retried_job = jobs[1]
+        self.assertEqual(retried_job.num_retries, 1)
+
+    @patch('data_refinery_foreman.surveyor.message_queue.app.send_task')
+    def test_retrying_lost_downloader_jobs(self, mock_send_task):
+        batch = self.insert_batch()
+        job = DownloaderJob.create_job_and_relationships(
+            num_retries=0, batches=[batch], downloader_task="dummy_task")
+        job.created_at = timezone.now() - main.MAX_QUEUE_TIME - timedelta(seconds=1)
+        job.save()
+
+        # Just run it once, not forever so get the function that is
+        # decorated with @do_forever
+        main.retry_lost_downloader_jobs.__wrapped__()
+        mock_send_task.assert_called_once()
+
+        jobs = DownloaderJob.objects.order_by('id')
+        original_job = jobs[0]
+        self.assertTrue(original_job.retried)
+        self.assertEqual(original_job.num_retries, 0)
+        self.assertFalse(original_job.success)
+
+        retried_job = jobs[1]
+        self.assertEqual(retried_job.num_retries, 1)
+
+    @patch('data_refinery_foreman.surveyor.message_queue.app.send_task')
+    def test_requeuing_processor_job(self, mock_send_task):
+        batch = self.insert_batch()
+        job = ProcessorJob.create_job_and_relationships(
+            num_retries=0, batches=[batch], pipeline_applied="AFFY_TO_PCL")
```

Comment:
Lines 165-167 are available in each test case of processor job. I wonder whether it is a good idea to put them in a separate function too.